api/net/Areas/Editor/Controllers/StorageController.cs

```diff
@@ -249,21 +249,45 @@ public IActionResult Stream([FromRoute] int? locationId, [FromQuery] string path
             if (dataLocation.Connection == null || dataLocation.Connection?.ConnectionType == ConnectionType.LocalVolume)
             {
                 var safePath = Path.Combine(_storageOptions.GetCapturePath(), path);
-                return GetResult(safePath, path);
+
+                return await GetResult(safePath, path);
             }
             else throw new NotImplementedException($"Location connection type '{dataLocation.Connection?.ConnectionType}' not implemented yet.");
         }
         else
         {
+            _logger.LogInformation("Getting stream for path: {Path}", path);
             var safePath = Path.Combine(_storageOptions.GetUploadPath(), path);
-            return GetResult(safePath, path);
+            return await GetResult(safePath, path);
         }
     }
 
-    private FileStreamResult GetResult(string safePath, string path)
+
+    private async Task<IActionResult> GetResult(string safePath, string path)
     {
-        if (!safePath.FileExists()) throw new NoContentException($"Stream does not exist: '{path}'");
+        _logger.LogInformation("Getting stream for path: {Path}", path);
+        var fileReference = await _fileReferenceService.GetByS3PathAsync(path);
+        _logger.LogInformation("File reference: {FileReference}", fileReference);
+        if (fileReference == null)
+        {
+            return NotFound($"Stream does not exist: '{path}'");
+        }
+
+        if (fileReference.IsSyncedToS3)
+        {
+            try
+            {
+                var stream = await _fileReferenceService.DownloadFromS3Async(path);
+                return File(stream, "application/octet-stream", fileReference.FileName);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "error on stream file from s3: {Path}", path);
```

Comment:
do you want to throw an exception here? otherwise, it will continue to the next statement 290.

Reply:
> do you want to throw an exception here? otherwise, it will continue to the next statement 290.

Hi Dao, this method try to find the path in s3 and then if s3 doesn't have, it will go to next step to check the local storage.

```        
if (!safePath.FileExists()) throw new NoContentException($"Stream does not exist: '{path}'");
```
both of them are not exist it will throw error, if it goes correct then it return file

 `return File(fileStream, info.MimeType!);`